### ISSUE TITLE
formatter: Should JSX syntax be supported with extensions other than `.jsx` and `.tsx`?

### ISSUE THREAD

**Issue by leaysgur** [oxc-project/oxc#14468]
Currently, `SourceType` is determined based on the file extension.

https://github.com/oxc-project/oxc/blob/0f19be09e9e827321e33cc534aee7e4da9fdc837/crates/oxc_formatter/src/service/source_type.rs

Therefore, while `.jsx` and `.tsx` files containing JSX syntax has no issue, but files with other extensions containing JSX syntax will result in parsing errors.

Should we enable JSX for non `.(j|t)sx` files?

Choices are:

- 1: Keep it as-is
- 2: Always enable JSX for JS source type only (like `oxlint`)
  - https://github.com/oxc-project/oxc/blob/0f19be09e9e827321e33cc534aee7e4da9fdc837/crates/oxc_linter/src/service/runtime.rs#L370
- 3: Always enable JSX for JS source type, and check source text to determine for TS source type (like Prettier)
  - https://github.com/prettier/prettier/blob/0d1e7abd5037a1fe8fbcf88a4d8cd13ec4d13a78/src/language-js/parse/utils/jsx-regexp.evaluate.js#L2

**Comment by overlookmotel** [comment]
I believe that there are examples of TS code where they are parsed differently depending on whether source type is plain TS or TSX (angle brackets meaning different things in the 2 cases).

So in that case at least, we need the user to specify source type to be able to parse correctly.

That's my memory anyway, and I don't have an example. I'm not sure how prettier manages to work around this problem. Or maybe for a formatter it doesn't matter, as it's the same tokens either way?

**Comment by camc314** [comment]
there's always the problem of ts, vs tsx files

The following is invalid TSX, as the `<T>` is interpreted as a jsx opening element. 
```tsx
const MyC = <T>() => {}
```

For it to be valid, a `,` must be added like so:
```tsx
const MyC = <T,>() => {}
```

That said, i don't believe this is a problem in JS files since type params don't exist there

**Comment by leaysgur** [comment]
Thank you!

> how prettier manages to work around this problem.

As I mentioned at the top comment, Prettier uses regexp to check for JSX-like strings in the TS source code before parsing.

However, doing this for all `.ts` files (except `.tsx` and `.d.ts`) is overhead.
So the issue is, should we follow that behavior too...?

For JS, since:

- Syntax doesn't conflict
- (It's not frequent) but still occasionally observed

I think it might be acceptable to enable JSX always like `oxlint` does.

**Comment by Dunqing** [comment]
I think we can enable JSX for JS files without any checks. `.js` files with JSX syntax are fairly common in old codebases, and it will also align with Prettier

**Comment by leaysgur** [comment]
👉🏻 #14605

- Always enable JSX for all JS source type
- Keep it as-is for TS source type for now
  - My local [benchmark](https://github.com/leaysgur/oxfmt-bench) shows about -10% perf just checking `source_text.contains("/>")`
  - May revisit in the future
